File: core/src/main/java/heroes/journey/entities/buffs/Buff.java
package heroes.journey.entities.buffs;

import heroes.journey.entities.tagging.Attributes;
import lombok.Builder;
import lombok.Getter;
import lombok.NonNull;

@Builder
@Getter
public class Buff {

    @NonNull
    private String name;
    @Builder.Default
    private int turnsBuffLasts = -1;
    @Builder.Default
    private int timesBuffCanBeUsed = -1;
    private final Attributes attributes = new Attributes();

    @Override
    public String toString() {
        return name;
    }

    public Buff register() {
        return BuffManager.register(this);
    }
}

File: modlib/src/main/java/heroes/journey/modlib/misc/Buff.kt
package heroes.journey.modlib.misc

import heroes.journey.modlib.attributes.Attributes
import heroes.journey.modlib.registries.Registrable
import heroes.journey.modlib.registries.Registries

/**
 * A Buff, used for temporary stat modifications.
 * This is a simple data container with no complex functions.
 */
class Buff(
    id: String,
    val turnsBuffLasts: Int,
    val attributes: Attributes
) : Registrable(id) {

    override fun register(): Buff {
        Registries.BuffManager.register(this)
        return this
    }
}

/**
 * Builder for defining a buff in a natural DSL style.
 */
class BuffBuilder {
    var id: String = ""
    var turnsBuffLasts: Int = 1
    private var _attributes: Attributes? = null

    fun attributes(init: heroes.journey.modlib.attributes.AttributesBuilder.() -> Unit) {
        _attributes = heroes.journey.modlib.attributes.attributes(init)
    }

    fun builtAttributes(): Attributes = _attributes ?: heroes.journey.modlib.attributes.attributes {}
}

/**
 * DSL entrypoint for defining a buff.
 *
 * Example usage:
 * ```kotlin
 * buff {
 *     id = "rested"
 *     turnsBuffLasts = 3
 *     attributes {
 *         stat(Ids.STAT_BODY, 2)
 *     }
 * }
 * ```
 */
fun buff(init: BuffBuilder.() -> Unit): Buff {
    val builder = BuffBuilder()
    builder.init()
    return Buff(builder.id, builder.turnsBuffLasts, builder.builtAttributes())
}
